### DISCUSSION TITLE
Disable battery-discharging on H3.AC3

### DISCUSSION THREAD

**Discussion by abuesen** [nathanmarlor/foxess_modbus#498] [Q&A] [answered]
Much thanks to the team that do this work. I like the easy way to connect to my FoxESS H3.AC3. Unfortuanly one cannot set the working-mode via modbus and I did not find a way to set it from HomeAssistant by the Fox-Cloud. The ususal work-mode is "Self-Use" at our site.

I want to supress discharging of the H3-battery if charging the electric car. Any solution or idea is welcome.

What about the registers 44000, 44001, 44002-44003 and 44004-44005? Did anyone already work with them?
Is there already a way to read the values in foxess_modbus?

My idea:
If wallbox.charging=enabled set *Remote Power Control Enable* (44000) to 'true' and set *Remote control-Active power command* (44002-44003) to '0'.
If wallbox.charging=false set *Remote Power Control Enable* (44000) to 'false'.

Any comments regarding this idea? 
I'm a little feared to test it without reading and displaying the values before.

Regards, Andre.

**Comment by canton7** [comment]
The H3 only exposes holding registers, which are not writable.

**Reply by abuesen** [reply]
So the documentation I got from FoxESS "H3/AIO-H3 Hybrid Inverter
MODBUS Protocol" date 2022-07-10 is wrong? There are these writable registers documented. I did not test it, yet.

**Reply by canton7** [reply]
In our testing, the H3 family only exposes holding registers and does not expose input registers.

If you are able to access input registers then fantastic! Maybe there was a firmware update which changed that.

**Reply by abuesen** [reply]
My firmware
Master : 1.83, Slave : 1.03, Manager : 1.66

Is there an easy way to add other holding registers to foxess_modbus? 
Then I'll try to read 44000 first and then try to write it with the service.

So I miss the 330xx and 33100, too. The registers hold firmware-information and *Modbus Protocol Version*

**Reply by canton7** [reply]
If you're testing, use the command-line modpoll tool.

**Reply by abuesen** [reply]
Ok, I have to work on this afternoon. I'm on duty now.
I got an error for now. Maybe the register is weired.

modpoll -m tcp -a 3 -1 -r 31000   192.168.100.7
Illegal Data Address exception response!

**Reply by canton7** [reply]
You might need to pass `-0` as well.

**Reply by abuesen** [reply]
The -0 was the key to read the right register.

But it did not work for every register I tried.

**Comment by abuesen** [comment]
It works. I used modpoll to see the values and foxess_modbus to set the register. (modpoll faild to write, assume my fault in handling and adressing)

So if I charge the EV and set 44000 to '1' the H3 stops working. Setting to '0' starts working again.
It really stops working complete, not only discharging the battery. It even stops generating power from PV.
A test with 44002 set to 2000 (Watts) results in generating power from PV.

I don't know how to set negative values. Register 44001-44002 uses negative values to represent discharge setting. But I don't understand the meaning yet.

Can you tell me how to set negative values via foxess_modbus service?

Next step is to build a simple control to regulate generated power regarding the PV-production.

**Reply by canton7** [reply]
Ooh, OK. That's exciting!

Can you see whether work mode / min/max soc / charge periods are writable?

**Reply by abuesen** [reply]
no, I think they are not. Just the registers mentioned above are documented. Maybe you can tell me other registers I can try?

**Reply by abuesen** [reply]
I did some basic testing. 

If *Remote power control* is disabled the H3 sets values in 44002-44003 and 44004-44005.
It might be interesting to observe this values. I was not able to get two registers as I32 with modpoll.

If *Remote power control* is enabled the H3 accepts values in 44002-44003 and 44004-44005. With positive values it generates power from battery (it is dark outside). Today I saw already generating AC-power from PV with the same values.

But I'm only able to set the first register and I'm only able to set values >=0.
Need help at this point.

PS: There are not output register holding information about workmode, min/max soc and charge periods at my H3.

**Reply by abuesen** [reply]
One more test, nice sun and SoC 32%.
If I Enable *Remote Power Control* the battery is still loading. With 44002 (positive values) I can set how many power is exportet by the inverter. So it does not regulate the needed load of the house, it just loads battery.

I assume yesterday it stops loading the battery because the SoC was 99%.

**Comment by abuesen** [comment]
Yeah! I found 41000 (*Work Mode*) and 41011 (*Minimum SoC on Grid*). It is possible to read and write.

This is the solution to my problem: Setting "Minimum SoC" to "SoC" or Workmode "Backup Only" if the EV is charging. 
That is much easier than calculating power myself.

This registers are undocumented for H3. I did not know them before. Found them in the documentation for KH in the Wiki.

**Reply by canton7** [reply]
Aha, interesting! We knew about the holding registers in the 41000 range (https://github.com/TonyM1958/HA-FoxESS-Modbus/wiki/Modbus-Map.xls), but didn't think they were writable. Can you read/write them as input registers? Does that include the charge period registers (41001-41006)? You normally have to write those as a block.

Can you access the other registers (31000 etc) as input registers?

There are quite a few people who will get very excited if we find out the H3 supports writes - it's probably the most-requested feature.

**Reply by abuesen** [reply]
The registers for charge period results in an error. Is the modpoll-call wrong? Did not try to write. See below.
(So I'm newbie to modbus, registers and how to write as a block and so on)

31000 are not writeable. Didn't try to write 11000. Why write Pv1-Voltage? What do you want me try to write?

41000 works fine. Setting it to '2' results in changing workmode in the cloud. I'm awaiting my wife to test it with the car.

Test with modpoll:
```
modpoll -m tcp -0 -a 3 -1 -r 41001 192.168.100.7
modpoll 3.10 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: MODBUS/TCP, FC3
Slave configuration...: address = 3, start reference = 41001 (PDU), count = 1
Communication.........: 192.168.100.7, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, output (holding) register table

-- Polling slave...
Illegal Data Address exception response!

```

**Reply by canton7** [reply]
I mean, is it possible to read 31000 as an input register? As in, `modpoll -m tcp -0 -a 3 -1 -t 3 -r 31000 192.168.100.7`.

(`-t 3` is "16-bit input register data type")

(On the H1 series, all registers are input registers, even ones which can't be written to. Being able to read all registers as the same type helps make the code a lot simpler.)

For 41001 and similar, those will definitely be input registers. Try: `modpoll -m tcp -0 -a 3 -1 -t 3 -r 41001 192.168.100.7`

**Reply by abuesen** [reply]
Both calls "time out".  Without `-t 3` the 31000 gives "4019" and 41001 "Illegal Data Adress".

I read the 41000  without `-t 3`! See below.

Wrong Tool? I use this one: https://www.modbusdriver.com/modpoll.html  on Windows.

```
modpoll -m tcp -0 -a 3 -1 -r 41000 192.168.100.7
modpoll 3.10 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: MODBUS/TCP, FC3
Slave configuration...: address = 3, start reference = 41000 (PDU), count = 1
Communication.........: 192.168.100.7, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, output (holding) register table

-- Polling slave...
[41000]: 0
```
```
modpoll -m tcp -0 -a 3 -1 -r -t 3 41000 192.168.100.7
modpoll: Illegal write data value! Try -h for help.
```

**Reply by abuesen** [reply]
The requested operations

```
modpoll -m tcp -0 -a 3 -1 -t 3 -r 31000 192.168.100.7
modpoll 3.10 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: MODBUS/TCP, FC4
Slave configuration...: address = 3, start reference = 31000 (PDU), count = 1
Communication.........: 192.168.100.7, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, input register table

-- Polling slave...
Reply time-out!

```

```
modpoll -m tcp -0 -a 3 -1  -r 31000 192.168.100.7
modpoll 3.10 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: MODBUS/TCP, FC3
Slave configuration...: address = 3, start reference = 31000 (PDU), count = 1
Communication.........: 192.168.100.7, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, output (holding) register table

-- Polling slave...
[31000]: 4019
```

**Reply by canton7** [reply]
Weird, so how did you write 41000? Can you write that using modpoll? For example, `modpoll -m tcp -0 -a 3 -t 4 -1 -r 41000 -c 1 192.168.100.7 0`

If so, can you write the charge periods using modpoll? Something like `modpoll -m tcp -a 3 -t 4 -r 41001 -0 -c 6 192.168.100.7 0 0 0 0 0 0` (that will disable all charge periods).

**Reply by abuesen** [reply]
> Weird, so how did you write 41000? Can you write that using modpoll? For example, `modpoll -m tcp -0 -a 3 -t 4 -1 -r 41000 -c 1 192.168.100.7 0`

"Time Out"
I write the values with the foxess_modbus service "Write Registers", works very well 😄 

> If so, can you write the charge periods using modpoll? Something like `modpoll -m tcp -a 3 -t 4 -r 41001 -0 -c 6 192.168.100.7 0 0 0 0 0 0` (that will disable all charge periods).

"Illegal Data Address exception response!"

**Reply by canton7** [reply]
That's odd, I wonder what the difference between modpoll and HA is there. Can you try writing 6 0's to register 41001 onwards with HA?

**Reply by abuesen** [reply]
Car arrived. The inverter is set to "Backup" automaticly if the meter for walboxes shows more than 500W. Very fine.

**Reply by canton7** [reply]
(Thanks a lot for your help here! I don't have an H3 to play with)

**Reply by abuesen** [reply]
(You're welcome. Tell me if I can give further assistance.)

**Reply by abuesen** [reply]
> That's odd, I wonder what the difference between modpoll and HA is there. Can you try writing 6 0's to register 41001 onwards with HA?

Raises an error.
Translation means: "Error while calling the service ...". It is the same error raised if I tried to write negative values to 44002.

![image](https://github.com/nathanmarlor/foxess_modbus/assets/136979340/7ff7323a-c322-4b8e-a265-894d1e8be0d9)

**Comment by abuesen** [comment, marked as answer]
Conclusion:

The undocumented register 41000 *Workmode* works on my H3. It is set to '2' if the consumption of the wallboxes is over 500W and reset to '0' if it is lower than 500W by HomeAssistant using the service "foxess_modbus: Write Registers". (Other registers are working, too)

This way the battery won't drain if the PV-power lowers or if we charge the car with full power.

Thanks to @canton7 for assistance!

**Comment by canton7** [comment]
Can you try the v1.8.2b1 pre-release please? I've:

1. Enabled the work mode, max charge current, max discharge current, max soc, min soc, min soc on grid sensors
2. Enable `number` entities for the above, which let you change their values
3. Enabled the charge period entities, and a service to check charge periods

I'm interested in:

1. Does it even start at all? You might need to look at the logs in Settings - System - Logs
2. Do those new sensors all read values?
3. Are you able to set the values of those new sensors?
4. Are you able to set charge periods using the Set Charge Period service?

Thanks!

**Comment by abuesen** [comment]
> 1. Enabled the work mode, max charge current, max discharge current, max soc, min soc, min soc on grid sensors

They do not show up. The logs show lots of errors. Send you the log as PM.

(Sorry I closed this issue because it was a question. Do you want to reopen or raise a new one?)

```
Error while executing automation automation.pubish_grid_power: Error rendering data template: ValueError: Template error: float got invalid input 'unknown' when rendering template '{{ (states('sensor.pv_power_fox') | float * 1000) | round(0) }}' but no default was specified
18:25:20 – (FEHLER) Automatisierung - Die Nachricht ist zum ersten Mal am 18:16:40 aufgetreten und erscheint 53 mal
Pubish Grid Power: Error executing script. Error for call_service at pos 3: Error rendering data template: ValueError: Template error: float got invalid input 'unknown' when rendering template '{{ (states('sensor.pv_power_fox') | float * 1000) | round(0) }}' but no default was specified
18:25:20 – (FEHLER) Automatisierung - Die Nachricht ist zum ersten Mal am 18:16:40 aufgetreten und erscheint 53 mal
Value (-3: -0.30000000000000004) for entity 'sensor.fox_pv1_current' address(es) '[31001]' failed validation against rule (Range : {'_min': 0, '_max': 100})
18:23:11 – (WARNUNG) FoxESS - Modbus (benutzerdefinierte Integration) - Die Nachricht ist zum ersten Mal am 18:07:33 aufgetreten und erscheint 19 mal
Autodetect: failed to connect to (tcp://192.168.100.7:502)
18:17:26 – (FEHLER) FoxESS - Modbus (benutzerdefinierte Integration)
TemplateError('ValueError: Template error: float got invalid input 'unavailable' when rendering template '{{ '%0.1f' | format(states('sensor.fox_inv_power_r') | float + states('sensor.fox_inv_power_s') | float + states('sensor.fox_inv_power_t') | float) }}' but no default was specified') while processing template 'Template<template=({{ '%0.1f' | format(states('sensor.fox_inv_power_r') | float + states('sensor.fox_inv_power_s') | float + states('sensor.fox_inv_power_t') | float) }}) renders=300>' for attribute '_attr_native_value' in entity 'sensor.inverter_power'
18:16:32 – (FEHLER) Template - Die Nachricht ist zum ersten Mal am 18:16:32 aufgetreten und erscheint 4 mal

```

**Reply by canton7** [reply]
Thanks for the logs. There don't seem to actually be any errors from this integration -- we're getting a negative current for pv1/pv2 (which I might correct), but that's unrelated.

So it looks like we're reading the work mode, max charge current, max discharge current, max soc, min soc and min soc on grid sensors. Can you confirm whether they're displaying the correct value, and whether you can change them?

**Reply by abuesen** [reply]
Ok, what I did:
Deleted integration, Update Repository with beta, Restart HA, Add Integration.

Before and after there are 62 entities, some disabled. I cannot find any additional "soc". Searched with ctrl-f. 

So what I did there are not the additional values showing up. Did I wrong?

**Reply by abuesen** [reply]
Just maybe a stupid question: 
Where did you enable the workMode, minSoc and so on?
Should it be _entity_descriptions.py_ line 1470 ff?  There is no _H3_SET_ in the code.

**Comment by abuesen** [comment]
I did a little research myself and altered the code on HA.
Result: Two new entities. `sensor.fox_min_soc_on_grid` and `number.fox_min_soc_on_grid`. The setting of "Min soc on grid" works.

I tested input-registers `input=41011` at both places but it gives no values.

'WorkMode' doesn't work as expected. The entitiy appears but is unavailiable. Did not test other ones.

(PS: git diff says, 1.8.1==1.8.2b1)

```
    # Sensor kept for back compat
    ModbusSensorDescription(
        key="min_soc_on_grid",
        addresses=[
            ModbusAddressesSpec(models=[*H1_SET, KH], input=[41011]),
            ModbusAddressesSpec(models=[*H3_SET], holding=[41011])
        ],
        name="Min SoC (On Grid)",
        device_class=SensorDeviceClass.BATTERY,
        state_class=SensorStateClass.MEASUREMENT,
        native_unit_of_measurement="%",
        icon="mdi:battery-arrow-down",
        validate=[Range(0, 100)],
    ),
    ModbusNumberDescription(
        key="min_soc_on_grid",
        address=[ModbusAddressSpec(models=[*H3_SET], holding=41011)],
        name="Min SoC (On Grid)",
        mode=NumberMode.BOX,
        native_min_value=10,
        native_max_value=100,
        native_step=1,
        native_unit_of_measurement="%",
        device_class=NumberDeviceClass.BATTERY,
        icon="mdi:battery-arrow-down",
        validate=[Range(0, 100)],
    ),

```

![image](https://github.com/nathanmarlor/foxess_modbus/assets/136979340/8c36853d-bc5c-4fc4-8c28-4a307a4be349)

**Reply by canton7** [reply]
Damn, I did mess up that beta release, sorry!
